### PR TITLE
Fixed callable statement index out of bounds error; Fixed parameter not defined error

### PR DESCRIPTION
**Pull request by tkyc**
For https://github.com/microsoft/mssql-jdbc/issues/1871
- Fixes callable statement index out of bounds
- Fixes callable statement parameter not defined error

### CODE REVIEW

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java [hunk 1]
```diff
@@ -50,6 +52,9 @@ public class CallableStatementTest extends AbstractTest {
     private static String getObjectLocalDateTimeProcedureName = RandomUtil.getIdentifier("CallableStatementTest_getObjectLocalDateTime_SP");
     private static String getObjectOffsetDateTimeProcedureName = RandomUtil.getIdentifier("CallableStatementTest_getObjectOffsetDateTime_SP");
     private static String procName = RandomUtil.getIdentifier("procedureTestCallableStatementSpPrepare");
+    private static String manyParamsTable = RandomUtil.getIdentifier("manyParam_Table");
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java:55]: nitpick: why not add AbstractSQLGenerator.escapeIdentifier() here so we don't have to do this in multiple places?

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java [hunk 2]
```diff
@@ -67,13 +72,64 @@ public static void setupTest() throws Exception {
...
+        String createLogin = "USE MASTER;CREATE LOGIN NewLogin WITH PASSWORD=N'P4ssword', " +
+                "DEFAULT_DATABASE = MASTER, DEFAULT_LANGUAGE = US_ENGLISH;ALTER LOGIN NewLogin ENABLE;";
+        String createUser = "USE MASTER;CREATE USER NewUser FOR LOGIN NewLogin WITH DEFAULT_SCHEMA = [DBO];";
+        String grantExecute = "GRANT EXECUTE ON " + AbstractSQLGenerator.escapeIdentifier(manyParamProc) + " TO NewUser;";
+
+        // Need to create a user with limited permissions in order to run through the code block we are testing
+        // The user created will execute sp_sproc_columns internally by the driver, which should not return all
+        // the column names as the user has limited permissions
+        try (Connection conn = PrepUtil.getConnection(connectionString)) {
+            Statement stmt = conn.createStatement();
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java:106]: leak, check eclipse, there are other leak warnings as well below

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java [hunk 3]
```diff
@@ -67,13 +72,64 @@ public static void setupTest() throws Exception {
...
+            createUserDefinedType();
+            createTableManyParams();
+            createProcedureManyParams();createGetObjectOffsetDateTimeProcedure(stmt);
+        }
+    }
+
+    // Test Needs more work to be configured to run on azureDB as there are slight differences
+    // between the regular SQL Server vs. azureDB
+    @Test
+    @Tag(Constants.xAzureSQLDB)
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java:93]: is this the same issue with AzureDW?

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java [hunk 4]
```diff
@@ -329,4 +385,49 @@ private static void createGetObjectOffsetDateTimeProcedure(Statement stmt) throw
...
+                + " AS INSERT INTO "
+                + AbstractSQLGenerator.escapeIdentifier(manyParamsTable) + " VALUES(@p1, @p2, @p3, @p4, @p5, @p6, @p7, @p8, @p9, @p10)";
+        try (Statement stmt = connection.createStatement()) {
+            stmt.execute(sql);
+        }
+    }
+
+    private static void createTableManyParams() throws SQLException {
+        String type = AbstractSQLGenerator.escapeIdentifier(manyParamUserDefinedType);
+        String sql = "create table " + AbstractSQLGenerator.escapeIdentifier(manyParamsTable) +
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java:411]: nitpick: you capitalized SQL in all other places can we be consistent here?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java [hunk 5]
```diff
@@ -1348,13 +1348,17 @@ private int findColumn(String columnName) throws SQLServerException {
...
+            parameterNamesSize = parameterNames.size();
         }
-        if (l == 0) { // Server didn't return anything, user might not have access
-            map.putIfAbsent(columnName, ai.incrementAndGet());
-            return map.get(columnName);// attempting to look up the first column will return no access exception
+
+        // @RETURN_VALUE will always be in the parameterNames map, so parameterNamesSize will always be at least of size 1.
+        // If the server didn't return anything (eg. the param names for the sproc), user might not have access.
+        // So, parameterNamesSize must be of size 1.
+        if (parameterNamesSize == 1) {
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCallableStatement.java:1360]:
```suggestion
        if (null != parameterNames && parameterNames.size() == 1) {
```
I think we can remove 1351 - 1355. parameterNamesSize only has 1 use, to see if its 1, so we don't really need it to be defined, and I think we can consolidate it use to just a line, without it looking too ugly.

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java [hunk 6]
```diff
@@ -368,6 +368,11 @@ public static void dropTypeIfExists(String typeName, java.sql.Statement stmt) th
         dropObjectIfExists(typeName, "TT", stmt);
     }
```
Comment (Jeffery-Wasty) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java:370]: Please add a simple javadoc.

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java [hunk 7]
```diff
@@ -368,6 +368,11 @@ public static void dropTypeIfExists(String typeName, java.sql.Statement stmt) th
...
 
+    public static void dropUserDefinedTypeIfExists(String typeName, Statement stmt) throws SQLException {
+        String sql = "IF EXISTS (select * from sys.types where name = '" + escapeSingleQuotes(typeName) + "') DROP TYPE " + typeName;
+        stmt.executeUpdate(sql);
```
Comment (Jeffery-Wasty) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java:373]: Can we just combine the lines here? Other functions in this file are combined. Just a nitpick tho, feel free to ignore.

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java [hunk 8]
```diff
@@ -281,52 +339,104 @@ public void inputParamsTest() throws SQLException {
...
+    }
+
+    private static void createUserDefinedType() throws SQLException {
+        String TVPCreateCmd = "CREATE TYPE " + manyParamUserDefinedType + " FROM MONEY";
+        try (Statement stmt = connection.createStatement()) {
+            stmt.executeUpdate(TVPCreateCmd);
+        }
+    }
+
+    private static void dropUserDefinedType() throws SQLException {
```
Comment (Jeffery-Wasty) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableStatementTest.java:436]: Unused?

**Review by Jeffery-Wasty — APPROVED** [review]

**Review by Jeffery-Wasty — APPROVED** [review]

**Review by lilgreenbird — APPROVED** [review]

**Pull request merged**